### ISSUE TITLE
MPI_ERR_COMM: invalid communicator , MPI_ERRORS_ARE_FATAL processes in this communicator will now abort

### ISSUE THREAD

**Issue by gizenmtl** [open-mpi/ompi#7270]
Hi everyone, 
I am new in MPI and I tried running my mpi codes but it gives error when I try to run it after compiling the file.
I run these commands
```
mpicc -g -Wall -o mm matrix2d.c -lm
mpiexec -n 6 mm
```
And then I got the error:
```
  [ubuntu:2511] *** An error occurred in MPI_Comm_rank
  [ubuntu:2511] *** reported by process [200015873,5]
  [ubuntu:2511] *** on communicator MPI_COMM_WORLD
  [ubuntu:2511] *** MPI_ERR_COMM: invalid communicator
  [ubuntu:2511] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
  [ubuntu:2511] ***    and potentially your MPI job)
```
My main function is below:
```
int main( int argc,  char *argv[]){
   int myrank, npes;
   MPI_Init (&argc, &argv);                //Initializations
   MPI_Comm comm = MPI_COMM_WORLD ; 
   MPI_Comm_rank (comm, &myrank);
   MPI_Comm_size (comm, &npes);  //MPI RANK,SIZE
   int n=5;
   double new_a;
   double new_b;
   double new_x;  
   MatrixVectorMultiply_2D(n, &new_a, &new_b, &new_x, comm);
   MPI_Finalize();
  return(0);
}
```
Other function parameters are:

```  
void MatrixVectorMultiply_2D(int n, double *a, double *b, double *x, MPI_Comm comm)
```
I'm just learning MPI and I'm stuck on this. Nobody around me could find a solution to my problem, so I found the remedy by consulting you. Could you please help me find out exactly what my problem is and how I can solve it?
Regards

**Comment by rhc54** [comment]
Which OMPI version are you using?

On Jan 3, 2020, at 11:40 AM, Gizen Mutlu <notifications@github.com <mailto:notifications@github.com> > wrote:

Hi everyone,
I am new in MPI and I tried running my mpi codes but it gives error when I try to run it after compiling the file.
I run these commands


mpicc -g -Wall -o mm matrix2d.c -lm
mpiexec -n 6 mm


And then I got the error:


  [ubuntu:2511] *** An error occurred in MPI_Comm_rank
  [ubuntu:2511] *** reported by process [200015873,5]
  [ubuntu:2511] *** on communicator MPI_COMM_WORLD
  [ubuntu:2511] *** MPI_ERR_COMM: invalid communicator
  [ubuntu:2511] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
  [ubuntu:2511] ***    and potentially your MPI job)


My main function is below:


int main( int argc,  char *argv[]){
   int myrank, npes;
   MPI_Init (&argc, &argv);                //Initializations
   MPI_Comm comm = MPI_COMM_WORLD ; 
   MPI_Comm_rank (comm, &myrank);
   MPI_Comm_size (comm, &npes);  //MPI RANK,SIZE
   int n=5;
   double new_a;
   double new_b;
   double new_x;  
   MatrixVectorMultiply_2D(n, &new_a, &new_b, &new_x, comm);
   MPI_Finalize();
  return(0);
}


Other function parameters are:


void MatrixVectorMultiply_2D(int n, double *a, double *b, double *x, MPI_Comm comm)


I'm just learning MPI and I'm stuck on this. Nobody around me could find a solution to my problem, so I found the remedy by consulting you. Could you please help me find out exactly what my problem is and how I can solve it?
Regards

**Comment by gizenmtl** [comment]
`$ ompi_info`
  Package: Open MPI buildd@lcy01-amd64-009 Distribution
  Open MPI: 2.1.1
  Open MPI repo revision: v2.1.0-100-ga2fdb5b
  Open MPI release date: May 10, 2017
   Open RTE: 2.1.1
  Open RTE repo revision: v2.1.0-100-ga2fdb5b
  Open RTE release date: May 10, 2017
   OPAL: 2.1.1
  OPAL repo revision: v2.1.0-100-ga2fdb5b
  OPAL release date: May 10, 2017
   MPI API: 3.1.0
    Ident string: 2.1.1
     Prefix: /usr
 Configured architecture: x86_64-pc-linux-gnu

**Comment by rhc54** [comment]
Pretty sure you cannot just set ```comm = MPI_COMM_WORLD``` as ```MPI_COMM_WORLD``` is a struct in OMPI, not an integer. Try just using ```MPI_COMM_WORLD``` in place of ```comm``` in your code.

**Comment by gizenmtl** [comment]
I tried your code since 3 hours but I must use `comm` variable cause it is used in my matrix-vector multiplication `MatrixVectorMultiply_2D(int n, double *a, double *b, double *x, MPI_Comm comm)` function by parameter. 
I'm sorry for asking, but I couldn't figure out how. How do I edit my code the way you say it?

**Comment by rhc54** [comment]
Like this:
```c
int main( int argc,  char *argv[]){
   int myrank, npes;
   MPI_Init (&argc, &argv);                //Initializations
   MPI_Comm_rank (MPI_COMM_WORLD, &myrank);
   MPI_Comm_size (MPI_COMM_WORLD, &npes);  //MPI RANK,SIZE
   int n=5;
   double new_a;
   double new_b;
   double new_x;
   MatrixVectorMultiply_2D(n, &new_a, &new_b, &new_x, MPI_COMM_WORLD);
   MPI_Finalize();
  return(0);
}
```

**Comment by gizenmtl** [comment]
I tried it but I got the same error :( 
Actually I tried like this :
```
int main( int argc,  char *argv[]){
   int myrank, npes;
   
   MPI_Init (&argc, &argv);                //Initializations
    
   MPI_Comm comm = MPI_COMM_WORLD ; 
   MPI_Comm_rank (MPI_COMM_WORLD, &myrank);
   MPI_Comm_size (MPI_COMM_WORLD, &npes);  //MPI RANK,SIZE
   
  /* int n=5;
   double new_a;
   double new_b;
   double new_x; */ 
   
   printf("i am here");
   
  // MatrixVectorMultiply_2D(n, &new_a, &new_b, &new_x,comm);
   MPI_Finalize();
   return(0);
}
```
It works. So, in this case,  is this my fault in these codes?
```
int n=5;
   double new_a;
   double new_b;
   double new_x;
   MatrixVectorMultiply_2D(n, &new_a, &new_b, &new_x, MPI_COMM_WORLD);
```

**Comment by rhc54** [comment]
Afraid I am not familiar with that function call as it isn't part of MPI. My guess is that you are indeed not using it correctly - you might need to look at the documentation for whatever package you found it in.

**Comment by gizenmtl** [comment]
I think my main problem is with the C code. There may be a communication error between the variables in the `MatrixVectorMultiply_2D` function and the variables in the main function.
Thank you four your attention :) I closed the issue.

**Comment by ggouaillardet** [comment]
This question has already been posted on Stack Overflow and it turns out the issue is in MatrixVectorMultiply_2D()

comm = MPI_COMM_WORLD

is a valid assignment since a MPI_Comm is implemented as a pointer to something (and that something is an opaque struct but that does not matter here)

Cheers,

Gilles

**Comment by bosilca** [comment]
To add to @ggouaillardet comment, the MPI standard explicitly mandates ___(Section 2.5.4 page 15 line 15)___ all MPI constants to be link-time constants, enabling their use in expressions and assignments. It should be noted that the MPI standard does not requires these constants to be compile-time constants, which means they can not necessarily be portably used as switch cases.

**Comment by ReyCTaylor** [comment]
Hi, everyone:
I am attempting to run paralell adcswan (padcswan) on a Linux virtual machine, but after preparing subdomanins and call the application I receive this error message "**An error occurred in MPI_Irecv reported by process [3643015169,0] on communicator MPI COMMUNICATOR 4 CREATE FROM 3 MPI_ERR_RANK: invalid rank MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort, and potentially your MPI job)**", and my job stops.
It is some problem with mpi library?
Please, help. Thank you, Reinaldo

**Comment by ReyCTaylor** [comment]
Hello, everybody. I am recalling over the error message **"An error occurred in MPI_Irecv reported by process [3375824897,0] on communicator MPI COMMUNICATOR 4 CREATE FROM 3 MPI_ERR_RANK: invalid rank MPI_ERRORS_ARE_FATAL"**
I am not programing with MPI, I am just using this feature installed with model ADCIRC in order to run ADCIRC or ADCIRC+SWAN in parallel. Please, any ADCIRC user or developer who had experienced this kind of error has found a solution to shared it?
I am running in Linux Ubuntu 20.04 (64-bits) mounted on Oracle VM VirtualBox 6.1, using 8 processors from 20 working for the main operative system. 
Thank you very much again, Reinaldo

**Comment by devreal** [comment]
@ReyCTaylor Please don't piggyback on old and already closed tickets. The problem you are seeing is very likely in ADCIRC, not in Open MPI. I recommend you get in touch with the ADCIRC developers to track down this issue. If you feel like this is *really* an issue in Open MPI (e.g., because ADCIRC works with MPICH with the exact same input configuration), please open a new ticket and provide all the information requested in the issue template (including the Open MPI config, hardware platform, libraries, etc)
